File: java/Exercises.java
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.FileNotFoundException;

public class Exercises {

    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    public static Optional<String> firstThenLowerCase(List<String> strings, Predicate<String> predicate) {
        return strings.stream()
                .filter(predicate)
                .findFirst()
                .map(String::toLowerCase);
    }

    static record Sayer(String phrase) {
        Sayer and(String word) {
            if (word == null || word.isEmpty()) {
                return this; // Return the current phrase unchanged if the word is null or empty
            }

            String trimmedWord = word.trim();

            // If the current phrase is empty, we return the trimmed word directly
            if (this.phrase.isEmpty()) {
                return new Sayer(trimmedWord);
            } else {
                // Append the trimmed word with a space only if the trimmedWord is not empty
                return new Sayer(this.phrase + (this.phrase.endsWith(" ") ? "" : " ") + trimmedWord);
            }
        }
    }

    public static Sayer say() {
        return new Sayer("");
    }

    public static Sayer say(String word) {
        return new Sayer(word != null ? word.trim() : "");
    }

    public static long meaningfulLineCount(String filename) throws FileNotFoundException, IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            return reader.lines()
                    .filter(line -> !line.trim().isEmpty() && line.trim().charAt(0) != '#')
                    .count();
        } catch (FileNotFoundException e) {
            throw new FileNotFoundException("No such file");
        }
    }

}

record Quaternion(double a, double b, double c, double d) {

    public final static Quaternion ZERO = new Quaternion(0, 0, 0, 0);
    public final static Quaternion I = new Quaternion(0, 1, 0, 0);
    public final static Quaternion J = new Quaternion(0, 0, 1, 0);
    public final static Quaternion K = new Quaternion(0, 0, 0, 1);

    public Quaternion {
        if (Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    Quaternion plus(Quaternion other) {
        return new Quaternion(a + other.a, b + other.b, c + other.c, d + other.d);
    }

    Quaternion times(Quaternion other) {
        return new Quaternion(
                a * other.a - b * other.b - c * other.c - d * other.d,
                a * other.b + b * other.a + c * other.d - d * other.c,
                a * other.c - b * other.d + c * other.a + d * other.b,
                a * other.d + b * other.c - c * other.b + d * other.a);
    }

    public Quaternion conjugate() {
        return new Quaternion(a, -b, -c, -d);
    }

    public List<Double> coefficients() {
        return List.of(a, b, c, d);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();

        // Handle 'a' part
        if (a != 0) {
            sb.append(a);
        }

        // Handle 'b' part
        if (b != 0) {
            if (sb.length() > 0) {
                sb.append(b > 0 ? "+" : ""); // Add '+' for positive numbers
            }
            sb.append(b == 1 ? "i" : (b == -1 ? "-i" : b + "i")); // Handle special cases for 1 and -1
        }

        // Handle 'c' part
        if (c != 0) {
            if (sb.length() > 0) {
                sb.append(c > 0 ? "+" : ""); // Add '+' for positive numbers
            }
            sb.append(c == 1 ? "j" : (c == -1 ? "-j" : c + "j")); // Handle special cases for 1 and -1
        }

        // Handle 'd' part
        if (d != 0) {
            if (sb.length() > 0) {
                sb.append(d > 0 ? "+" : ""); // Add '+' for positive numbers
            }
            sb.append(d == 1 ? "k" : (d == -1 ? "-k" : d + "k")); // Handle special cases for 1 and -1
        }

        // If all coefficients are zero
        if (sb.length() == 0) {
            return "0";
        }

        return sb.toString();
    }

}

sealed interface BinarySearchTree permits Empty, Node {
    int size();

    boolean contains(String value);

    BinarySearchTree insert(String value);
}

final record Empty() implements BinarySearchTree {
    @Override
    public int size() {
        return 0;
    }

    @Override
    public boolean contains(String value) {
        return false;
    }

    @Override
    public BinarySearchTree insert(String value) {
        return new Node(value, this, this);
    }

    @Override
    public String toString() {
        return "()";
    }
}

final class Node implements BinarySearchTree {
    private final String value;
    private final BinarySearchTree left;
    private final BinarySearchTree right;

    Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    @Override
    public int size() {
        return 1 + left.size() + right.size();
    }

    @Override
    public boolean contains(String value) {
        return this.value.equals(value) || left.contains(value) || right.contains(value);
    }

    @Override
    public BinarySearchTree insert(String value) {
        if (value.compareTo(this.value) < 0) {
            return new Node(this.value, left.insert(value), right);
        } else {
            return new Node(this.value, left, right.insert(value));
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("(");
        if (!(left instanceof Empty)) {
            sb.append(left.toString());
        }
        sb.append(value);
        if (!(right instanceof Empty)) {
            sb.append(right.toString());
        }
        sb.append(")");
        return sb.toString();
    }

}

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

// Write your first then lower case function here
fun firstThenLowerCase(strings: List<String>, predicate: (String) -> Boolean): String? {
    for (string in strings) {
        if (predicate(string)) {
            return string.lowercase()
             }
    }
    return null
}


// Write your say function here
class Say(private val words: List<String> = listOf()) {
    fun and(word: String): Say {
        return Say(words + word)
    }
    val phrase: String
        get() = words.joinToString(" ")
}
fun say(word: String = ""): Say {
    return Say(listOf(word))
}


// Write your meaningfulLineCount function here
fun meaningfulLineCount(filename: String): Long {
    BufferedReader(FileReader(filename)).use { reader ->
        return reader.lines()
            .filter { line -> 
                line.isNotBlank() && !line.trimStart().startsWith("#")
            }
            .count() 
    }
}

// Write your Quaternion data class here
data class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double) {
    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }
    fun coefficients(): List<Double> = listOf(a, b, c, d)
    fun conjugate(): Quaternion {
        return Quaternion(a, -b, -c, -d)
    }
    operator fun plus(other: Quaternion): Quaternion {
        return Quaternion(
        a + other.a,
            b + other.b,
            c + other.c,
            d + other.d
        )
    }
    operator fun times(other: Quaternion): Quaternion {
        val valA = a * other.a - b * other.b - c * other.c - d * other.d
        val valB = a * other.b + b * other.a + c * other.d - d * other.c
        val valC = a * other.c - b * other.d + c * other.a + d * other.b
        val valD = a * other.d + b * other.c - c * other.b + d * other.a
        return Quaternion(valA, valB, valC, valD)
    }

    override fun toString(): String {
        val result = StringBuilder()

        if (a != 0.0) result.append(a)

        if (b != 0.0) {
            if (b > 0 && result.isNotEmpty()) result.append("+")
            result.append(formatting(b, "i"))
        }

        if (c != 0.0) {
            if (c > 0 && result.isNotEmpty()) result.append("+")
            result.append(formatting(c, "j"))
        }

        if (d != 0.0) {
            if (d > 0 && result.isNotEmpty()) result.append("+")
            result.append(formatting(d, "k"))
        }

        return if (result.isEmpty()) "0" else result.toString()
    }
    private fun formatting(value: Double, symbol: String): String {
        return when (value) {
            1.0 -> symbol
            -1.0 -> "-$symbol"
            else -> "$value$symbol"
        }
    }
}
// Write your Binary Search Tree interface and implementing classes here
sealed interface BinarySearchTree {
    fun insert(value: String): BinarySearchTree
    fun contains(value: String): Boolean
    fun size(): Int
    override fun toString(): String
    object Empty : BinarySearchTree {
        override fun insert(value: String): BinarySearchTree {
            return Node(value, Empty, Empty)
        }
        override fun contains(value: String): Boolean = false
        override fun size(): Int = 0
        override fun toString(): String = "()"
    }

    data class Node(
        val nodeValue: String,
        val leftChild: BinarySearchTree = Empty,
        val rightChild: BinarySearchTree = Empty
    ) : BinarySearchTree {
        override fun insert(newValue: String): BinarySearchTree {
            return if (newValue < nodeValue) {
                Node(nodeValue, leftChild.insert(newValue), rightChild)
            } else {
                Node(nodeValue, leftChild, rightChild.insert(newValue))
            }
        }
        override fun contains(searchValue: String): Boolean {
            return when {
                searchValue == nodeValue -> true
                searchValue < nodeValue -> leftChild.contains(searchValue)
                else -> rightChild.contains(searchValue)
            }
        }
        override fun size(): Int = 1 + leftChild.size() + rightChild.size()

        override fun toString(): String {
            val leftString = if (leftChild is Empty) "" else "$leftChild"
            val rightString = if (rightChild is Empty) "" else "$rightChild"
            return "($leftString$nodeValue$rightString)"
        }
    }
}

